compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -987,6 +1020,55 @@ impl<'context> Elaborator<'context> {
         self.generics.clear();
     }
 
+    fn check_parent_traits_are_implemented(&mut self, trait_impl: &UnresolvedTraitImpl) {
+        let Some(trait_id) = trait_impl.trait_id else {
+            return;
+        };
+
+        let Some(object_type) = &trait_impl.resolved_object_type else {
+            return;
+        };
+
+        let Some(the_trait) = self.interner.try_get_trait(trait_id) else {
+            return;
+        };
+
+        // Note: we only check if the immediate parents are implemented, we don't check recursively.
+        // Why? If a parent isn't implemented, we get an error. If a parent is implemented, we'll
+        // do the same check for the parent, so this trait's parents parents will be checked, so the
+        // recursion is guaranteed.
+        for parent_trait_bound in the_trait.trait_bounds.clone() {
+            let Some(parent_trait) = self.interner.try_get_trait(parent_trait_bound.trait_id)
+            else {
+                continue;
+            };
+
+            if self
+                .interner
+                .try_lookup_trait_implementation(
+                    object_type,
+                    parent_trait_bound.trait_id,
+                    &parent_trait_bound.trait_generics.ordered,
+                    &parent_trait_bound.trait_generics.named,
+                )
+                .is_err()
```

Comment:
Looks like we're using the original generics from the supertype here. Does this still work if they're changed in the program? E.g.

```rs
trait Foo<A> {}

trait Bar<B>: Foo<B> {}

impl Foo<i32> for () {}

impl Bar<i32> for () {}
```
Trying this I get `error: The trait Foo<<B>> is not implemented for ()`

Reply:
Ah, you are right.

Where do I get the type bindings from the `Trait`? I see there's `associated_types` and `generics` on `Trait`, but I'm not sure how to do it...

Reply:
I pushed a commit that solves this case... is there an example with associated types? (I don't fully know how that works in Rust nor in Noir yet)

Reply:
You can see if the following passes for associated types:
```rs
trait Foo { type A; }

trait Bar<B>: Foo<A = B> {}

impl Foo for () { type A = i32; }

impl Bar<i32> for () {}
```

Reply:
Apparently yes. I added a test for it.